### PR TITLE
remove some of the python2 compat and update syntax

### PR DESCRIPTION
**Pull request by danieleades**

### CODE REVIEW

File: poetry/core/masonry/builders/builder.py [hunk 1]
```diff
@@ -364,17 +363,12 @@ def convert_author(cls, author: str) -> Dict[str, str]:
     @classmethod
     @contextmanager
     def temporary_directory(cls, *args: Any, **kwargs: Any) -> ContextManager[str]:
```
Comment (neersighted) [inline poetry/core/masonry/builders/builder.py:365]: This is an example of a now-unnecessary wrapper function, and I would prefer to leave it in place on the first commit and then excise it entirely on a follow-up commit

File: poetry/core/masonry/builders/builder.py [hunk 2]
```diff
@@ -364,17 +363,12 @@ def convert_author(cls, author: str) -> Dict[str, str]:
...
-                yield name
-        except ImportError:
-            name = tempfile.mkdtemp(*args, **kwargs)
+        from tempfile import TemporaryDirectory
 
+        with TemporaryDirectory(*args, **kwargs) as name:
             yield name
 
-            shutil.rmtree(name)
+        shutil.rmtree(name)
```
Comment (neersighted) [inline poetry/core/masonry/builders/builder.py:371]: This change is strictly speaking incorrect -- the TemporaryDirectory contextmanager will clean up on its own. Removal of the entire wrapper is better than removing this line however.

**Review by neersighted — CHANGES_REQUESTED** [review]
Looks good overall with one exception -- the refactor of the TemporaryDirectory wrapper is incorrect.

**Comment by danieleades** [comment]
is there a reason not to simply add the pyupgrade hook?

**Comment by branchvincent** [comment]
https://github.com/python-poetry/poetry/pull/4758 adds the hook for `poetry`, so i imagine we should do the same here

**Comment by neersighted** [comment]
> 
> 
> [python-poetry/poetry#4758](https://github.com/python-poetry/poetry/pull/4758) adds the hook for `poetry`, so i imagine we should do the same here

The intention is to copy the automation from poetry once I've got it dialled in to everyone's satisfaction. Still, a PR adding pyupgrade would not be unwelcome -- I just plan to replace the `.pre-commit-config.yaml` in this repo in a day or two.

**Review by neersighted — DISMISSED** [review]

File: poetry/core/masonry/builders/builder.py (hunk 2 above)
Comment (neersighted) [reply poetry/core/masonry/builders/builder.py:371]: It looks like `rmtree` line is still around.

File: poetry/core/masonry/builders/builder.py (hunk 1 above)
Comment (neersighted) [inline poetry/core/masonry/builders/builder.py:365]: We should completely eliminate this `temporary_directory` method in the first commit.

**Review by neersighted — CHANGES_REQUESTED** [review]

File: poetry/core/masonry/builders/builder.py (hunk 1 above)
Comment (danieleades) [reply poetry/core/masonry/builders/builder.py]: that's reared it's ugly head again after a rebase. expunged again now

**Review by neersighted — APPROVED** [review]

**Pull request merged**